internal/controller/checkpointrestoreoperator_controller.go

```diff
@@ -263,41 +444,48 @@ func handleWriteFinished(ctx context.Context, event fsnotify.Event) {
 				"file",
 				c,
 			)
+			continue
 		}
 		archivesToDelete[fi.ModTime().UnixMicro()] = c
 	}
 
-	keys := make([]int64, len(archivesToDelete))
-	i := 0
-	for key := range archivesToDelete {
-		keys[i] = key
-		i++
+	// Handle excess checkpoints by count
+	if maxCheckpoints > 0 && checkpointArchivesCounter > maxCheckpoints {
+		excessCount := int64(checkpointArchivesCounter - maxCheckpoints)
+		log.Info("Checkpoint count exceeds limit", "checkpointArchivesCounter", checkpointArchivesCounter, "maxCheckpoints", maxCheckpoints, "excessCount", excessCount)
+		toDelete := selectArchivesToDelete(checkpointArchives, excessCount, ByCount)
+		for _, archive := range toDelete {
+			log.Info("Deleting checkpoint archive due to excess count", "archive", archive)
+			err := os.Remove(archive)
+			if err != nil {
+				log.Error(err, "removal of checkpoint archive failed", "archive", archive)
+			}
+			checkpointArchivesCounter--
+			if checkpointArchivesCounter <= maxCheckpoints {
+				break
+			}
+		}
 	}
-	sort.Slice(keys, func(i, j int) bool { return keys[i] < keys[j] })
-	for _, key := range keys {
-		log.Info(
-			"Deleting container checkpoint archive",
-			"container", details.container,
-			"archive", archivesToDelete[key],
-		)
+}
 
-		err = os.Remove(archivesToDelete[key])
-		if err != nil {
-			// If the removal fails the checkpoint archive might have been
-			// deleted externally. This is not fatal but unexpected.
-			// Logging it as an error, but not returning an error.
-			log.Error(
-				err,
-				"Removal of checkpoint archive failed",
-				"container", details.container,
-				"archive", archivesToDelete[key],
-			)
-		}
-		checkpointArchivesCounter -= 1
-		if checkpointArchivesCounter <= maxCheckpointsPerContainer {
-			break
+func selectArchivesToDelete(archives []string, excess int64, policy RetentionPolicy) []string {
+	toDelete := make([]string, 0)
+
+	switch policy {
+	case ByCount:
+		// Sort by modification time (oldest first)
+		sort.Slice(archives, func(i, j int) bool {
+			fi1, _ := os.Stat(archives[i])
```

Comment:
`fi1` is a little too terse to understand what refers too. How about using `fileInfo` instead?

Also, I'm a bit sceptical about ignoring the error here. It would be better to return false if either of the calls do not succeed, and print the error.